File: src/mod11_gringotts_concurrency/lesson8_phaser/java/Main.java
package mod11_gringotts_concurrency.lesson8_phaser.java;

import java.util.concurrent.Phaser;

public class Main {
    public static void main(String[] args) {

        BankAccount ollivander = new BankAccount("Ollivander's account", 15_000);
        BankAccount hagrid = new BankAccount("Hadrid's account", 500_000);

        Phaser phaser = new Phaser();

        Thread1 thread1 = new Thread1(hagrid, ollivander, phaser);
        var runnable = new RunnableForThread2(hagrid, ollivander, phaser);
        int ph2 = runnable.registerThreadInPhaser();
        Thread thread2 = new Thread(runnable);
        int ph1 = thread1.registerThreadInPhaser();

        thread1.start();
        thread2.start();

        phaser.awaitAdvance(ph1);//join()
        phaser.awaitAdvance(ph2);//join()

        System.out.println("Balances after transfers:" + hagrid.getAccountName() + ":" + hagrid.getBalance() + ";" +
                ollivander.getAccountName() + ":" + ollivander.getBalance());

    }
}

File: src/mod11_gringotts_concurrency/lesson8_phaser/kotlin/Main.kt
package mod11_gringotts_concurrency.lesson8_phaser.kotlin

import java.util.concurrent.CyclicBarrier
import java.util.concurrent.Exchanger
import java.util.concurrent.Phaser
import kotlin.concurrent.thread

fun main() {

    val hagridAccount = BankAccount("Hadrid's account", 500_000)
    val ollivanderAccount = BankAccount("Ollivander's account", 15_000)

    val phaser=Phaser()
    phaser.register()

    val thread1 = thread {
        phaser.register()
        repeat(100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
        }
        phaser.arriveAndAwaitAdvance()
    }
    val thread2 = thread {
        phaser.register()
        repeat(100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
        }
        phaser.arriveAndAwaitAdvance()
    }

    phaser.arriveAndAwaitAdvance()

    println(
        "Balances after transfers:" + hagridAccount.accountName + ":" + hagridAccount.balance + ";" +
                ollivanderAccount.accountName + ":" + ollivanderAccount.balance
    )


}